pywbem_mock/_mainprovider.py

```diff
@@ -0,0 +1,4209 @@
+
+    def Associators(self, namespace, ObjectName, AssocClass=None,
+                    ResultClass=None,
+                    Role=None, ResultRole=None, IncludeQualifiers=None,
+                    IncludeClassOrigin=None, PropertyList=None):
+        """
+        Implements a WBEM server responder for
+        :meth:`~pywbem.WBEMConnection.Associators`.
+
+        Parameters:
+
+          namespace (:term:`string`):
+            The name of the CIM namespace in the CIM repository (case
+            insensitive). Must not be `None`. Leading or trailing slash
+            characters are ignored.
+
+          ObjectName:
+            The object path of the source object:
+
+            * Instance-level use: The instance path of the
+              source instance, as a :class:`~pywbem.CIMInstanceName` object. If
+              this object specifies a namespace, it must match the namespace
+              parameter. Its `host` attribute will be ignored.
+
+            * Class-level use: The class path of the source
+              class, as a :term:`string`. The string is interpreted as a class
+              name in the default namespace of the connection
+              (case independent).
+
+          See :meth:`~pywbem.WBEMConnection.Associators` for description of
+          remaining parameters
+
+        Returns:
+
+            * For instance-level requests
+              (ObjectName is :class:`~pywbem.CIMInstanceName`): A list of
+              :class:`~pywbem.CIMInstance` objects that are representations
+              of the associated instances.
+
+              The `path` attribute of each :class:`~pywbem.CIMInstance`
+              object is a :class:`~pywbem.CIMInstanceName` object with its
+              attributes set as follows:
+
+              * `classname`: Name of the creation class of the instance.
+              * `keybindings`: Keybindings of the instance.
+              * `namespace`: Name of the CIM namespace containing the instance.
+              * `host`: Host and optionally port of the WBEM server containing
+                the CIM namespace.
+
+            * For class-level requests (ObjectName is :term:`string`):
+              A list of :func:`py:tuple` of (classpath, class) objects that are
+              representations of the associated classes.
+
+              Each tuple represents one class and has these items:
+
+              * classpath (:class:`~pywbem.CIMClassName`): The class
+                path of the class, with its attributes set as follows:
+
+                * `classname`: Name of the class.
+                * `namespace`: Name of the CIM namespace containing the class.
+                * `host`: Host and optionally port of the WBEM server containing
+                  the CIM namespace.
+
+              * class (:class:`~pywbem.CIMClass`): The representation of the
+                class, with its `path` attribute set to the `classpath` tuple
+                item.
```

Comment:
Proposal: Should we split up instance-level and class-level use into two provider calls? If we ever allow user-defined association providers, every user would be thankful of that.

Reply:
Let me answer that one after I get the prototype DefaultInstanceProvider done tomorrow.

Reply:
Pro:

1. Clarifies differences between class and instance

2. Andy notes that there are cases in IBM products that hide the instances of association classes.

3. Karl noted that causes 8 new methods in the providers.

CONCLUSION: We can do this later if we need it.